atomicapp/providers/openshift.py

```diff
@@ -36,91 +35,205 @@ class OpenShiftProvider(Provider):
+                                                  " name in artifacts metadata "
+                                                  "artifact=%s" % artifact)
+
+                url = self._get_url(namespace, kind, name)
+
+                if self.dryrun:
+                    logger.info("DRY-RUN: %s", url)
+                    continue
+
+                res = requests.delete(url, verify=self.ssl_verify)
+                if res.status_code == 200:
+                    logger.info(" %s sucessfully undeployed.", res.content)
+                else:
+                    msg = "%s %s" % (res.status_code, res.content)
+                    logger.error(msg)
+                    raise ProviderFailedException(msg)
+
+    def _process_artifacts(self):
+        """  parse artifact files
+
+        save parsed artifacts to self.openshift_artifacts
+        """
         for artifact in self.artifacts:
+            logger.debug("Procesesing artifact: %s", artifact)
             data = None
-            artifact_path = os.path.join(self.path, artifact)
-            with open(artifact_path, "r") as fp:
+            with open(os.path.join(self.path, artifact), "r") as fp:
                 data = anymarkup.parse(fp, force_types=None)
-            if "kind" in data:
-                if data["kind"].lower() == "template":
-                    logger.info("Processing template")
-                    artifact = self._processTemplate(artifact_path)
-                kube_order[data["kind"].lower()] = artifact
+                # kind has to be specified in artifact
+                if "kind" not in data.keys():
+                    raise ProviderFailedException(
+                        "Error processing %s artifact. There is no kind" %
+                        artifact)
+
+                kind = data["kind"].lower()
+
+                # process templates
+                if kind == "template":
+                    processed_objects = self._process_template(data)
+                    # add all processed object to artifacts dict
+                    for obj in processed_objects:
+                        obj_kind = obj["kind"].lower()
+                        if obj_kind not in self.openshift_artifacts.keys():
+                            self.openshift_artifacts[obj_kind] = []
+                        self.openshift_artifacts[obj_kind].append(obj)
+                    continue
+
+                # add parsed artifact to dict
+                if kind not in self.openshift_artifacts.keys():
+                    self.openshift_artifacts[kind] = []
+                self.openshift_artifacts[kind].append(data)
+
+    def _process_template(self, template):
+        """ process tempalate
+
+        return list of templated object from template
+        """
+        logger.debug("processing template: %s", template)
+        url = self._get_url(self._get_namespace(template), 'processedtemplates')
+        res = requests.post(url, json=template, verify=self.ssl_verify)
```

Comment:
In returns "rendered" template according to its parameters. Main reason for this is that OpenShift supports generating values from expression for parameters (https://docs.openshift.com/enterprise/3.0/architecture/core_concepts/templates.html)

It is a bit confusing because now we have two levels of templating :-( 
Example of this is here: https://github.com/kadel/openshift3mlbparks/blob/nulecule-two-templates/mlbparks-app-template.json
All `$app_*` variables are Nulecule parameters rest (uppercase variables) are OpenShifts